chain/store.go

```diff
@@ -364,3 +373,8 @@ func (store *Store) GenesisCid() cid.Cid {
 func (store *Store) Stop() {
 	store.headEvents.Shutdown()
 }
+
+// Status returns the current chain status.
```

Comment:
The Status method exists on the store for testing the store sets the status correctly on `SetHead` calls.

> The syncer is the only caller of SetHead

The chain store `Load()` method also calls `SetHead()`, and since the Syncer is instantiated before the store is loaded we needed a way to update the status once the store was done loading. The calls in the syncer code were extraneous, will remove those.

Reply:
@ZenGround0 are you alright leaving this here given the above?

Reply:
The blocking stands. I would prefer an undef tipset in heaviest head until there is a sync over adding an exported `Status` method to the store for the reasons stated above.

Reply:
An earlier implementation of this did just that (set values to undef), reviewers commented (https://github.com/filecoin-project/go-filecoin/pull/3305#discussion_r319777238) that we could get this information from the store. The way to address this was to share status between the syncer and the store. 

Here are the options I think we have to remove this blocking comment:
- accept this exported method as a necessary evil of testing
- create a new test file in the chain package (not in chain_test)
- define an exported method that accepts a chain store and returns its reporter
- instantiate a syncer in each store test and inspect its status return
- not test the store updates its status

I would also like to point out that `Status` has not been added to the store interface definitions, it's a method on the struct, nothing that accepts a store has access to it.

Reply:
Thanks for this context.

My takeaway from @anorth's comment is "if its easy to set the head at setup, then do it", not "the head must be set"

I do not accept that the existing approach is a necessary evil of testing.  Convenient coupling of the syncer and the store and adding to the store's exports is often attractive in the short term but in my experience it is the wrong thing to do.  

In my view we have two better different paths forward, preferable to this coupling.  
(A) we go back to your earlier implementation and leave the head value unset
(B) we introduce a `Start`/`StartStatus` function into the syncer that simply sets the Status head/head-height to a value determined by making a call to the chain store post load in `node.Start()`

> I would also like to point out that Status has not been added to the store interface definitions, it's a method on the struct, nothing that accepts a store has access to it.

This is true, thanks for flagging my imprecise language.  I am talking about the store's exported function set which I believe should follow guidelines for good API design even though go makes a point of taking selective subsets of it.

Reply:
> (A) we go back to your earlier implementation and leave the head value unset

We can do that, but it feels like a step backwards.

> (B) we introduce a Start/StartStatus function into the syncer that simply sets the Status head/head-height to a value determined by making a call to the chain store post load in node.Start()

I don't see how adding a new exported method to the syncer in place of one on the store is any simpler.

To clarify, is this blocking comment about the Store having an exported `Status` method or about the Store and Syncer sharing a status reporter?

Reply:
> I don't see how adding a new exported method to the syncer in place of one on the store is any simpler.

From my perspective it is a great deal simpler.  If we remove the shared dependency we remove all consideration of the syncer's status reporter when trying to understand the chain store.  This keeps the chain store's definition better contained; now we don't need to understand the syncing status code and node building/starting code to understand why the chain store is the way it is.  We also address the original concern of not having more methods that do the same thing exported from the store.

Yes we would be adding a method to the syncer in proposal B, and it is a well-scoped and simple method that does what we want without relying on sharing state with separate parts of the system.  It also does a new unique thing for the syncer that its other methods don't already do in some way.

> To clarify, is this blocking comment about the Store having an exported Status method or about the Store and Syncer sharing a status reporter?

At this point they seem roughly equivalent.  I originally blocked because the `Status` method seems like a poor fit for the store.  It has become clearer to me that you needed to add this method to the store as a result of adding functionality to the store that I believe the syncer/its caller should perform instead.  This strengthens my intuition that the method doesn't belong and suggests that the right fix is to remove the shared state.

Reply:
I don't like the idea of allowing code external to the syncer to influence it's state, from my understanding option B allows that.

In the interest of getting this merged I will revert the changes that allow the store to modify status and propose that in a follow on.